src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/SnapshotTracker.kt

```diff
@@ -0,0 +1,255 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2024 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck.strategy.managed
+
+import org.jetbrains.kotlinx.lincheck.findField
+import org.jetbrains.kotlinx.lincheck.isPrimitiveWrapper
+import org.jetbrains.kotlinx.lincheck.readFieldSafely
+import org.jetbrains.kotlinx.lincheck.strategy.managed.SnapshotTracker.MemoryNode.*
+import org.jetbrains.kotlinx.lincheck.util.*
+import java.lang.Class
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier
+import java.util.Collections
+import java.util.IdentityHashMap
+import java.util.Stack
+import java.util.concurrent.atomic.AtomicIntegerArray
+import java.util.concurrent.atomic.AtomicLongArray
+import java.util.concurrent.atomic.AtomicReferenceArray
+
+
+/**
+ * Manages a snapshot of the global static state.
+ *
+ * This class only tracks static memory and memory reachable from it,
+ * referenced from the test code. So the whole static state is not recorded, but only a subset of that,
+ * which is named *snapshot*.
+ */
+class SnapshotTracker {
```

Comment:
Should it be `SnapshotTracker` or `MemoryChangesTracker`?

Reply:
This should be tracker, because its methods are invoked even if no chnages to static memory were made (a read, for example, which we must track, but there is no immidiate change in memory)